kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,177 @@
+use std::sync::Arc;
+
+use crate::action_reconciliation::RetentionCalculator;
+use crate::log_segment::LogSegment;
+use crate::path::ParsedLogPath;
+use crate::snapshot::Snapshot;
+use crate::table_properties::TableProperties;
+use crate::{DeltaResult, Engine, Error, Version};
+
+use url::Url;
+
+use super::COMPACTION_ACTIONS_SCHEMA;
+use crate::checkpoint::log_replay::{CheckpointBatch, CheckpointLogReplayProcessor};
+use crate::engine_data::FilteredEngineData;
+use crate::log_replay::LogReplayProcessor;
+
+/// Determine if log compaction should be performed based on the commit version and
+/// compaction interval.
+pub fn should_compact(commit_version: Version, compaction_interval: Version) -> bool {
+    // Commits start at 0, so we add one to the commit version to check if we've hit the interval
+    compaction_interval > 0
+        && commit_version > 0
+        && ((commit_version + 1) % compaction_interval) == 0
+}
+
+/// Writer for log compaction files
+///
+/// This writer provides an API for creating log compaction files that aggregate actions
+/// from multiple commit files.
+#[derive(Debug)]
+pub struct LogCompactionWriter {
+    /// Reference to the snapshot of the table being compacted
+    snapshot: Arc<Snapshot>,
+    start_version: Version,
+    end_version: Version,
+    /// Cached compaction file path
+    compaction_path: Url,
+}
+
+impl RetentionCalculator for LogCompactionWriter {
+    fn table_properties(&self) -> &TableProperties {
+        self.snapshot.table_properties()
+    }
+}
+
+impl LogCompactionWriter {
+    pub fn try_new(
+        snapshot: Arc<Snapshot>,
+        start_version: Version,
+        end_version: Version,
+    ) -> DeltaResult<Self> {
+        if start_version > end_version {
+            return Err(Error::generic(format!(
+                "Invalid version range: start_version {} > end_version {}",
+                start_version, end_version
+            )));
+        }
+
+        // Compute the compaction path once during construction
+        let compaction_path =
+            ParsedLogPath::new_log_compaction(snapshot.table_root(), start_version, end_version)?;
+
+        Ok(Self {
+            snapshot,
+            start_version,
+            end_version,
+            compaction_path: compaction_path.location,
+        })
+    }
+
+    /// Get the path where the compaction file will be written
+    pub fn compaction_path(&self) -> &Url {
+        &self.compaction_path
+    }
+
+    /// Get an iterator over the compaction data to be written
+    ///
+    /// Performs action reconciliation for the version range specified in the constructor.
+    /// It creates a LogSegment that only includes commits in the specified version range,
+    /// and performs action reconciliation.
```

Comment:
nit: `LogSegment` is a private concept (or rather `internal-api`) so should probably make this part `//` comment instead of `///` docstring?